Develop var len obs feature (#4909)
Unity-Technologies/ml-agents: docs/Learning-Environment-Design-Agents.md

- Use as few rays and tags as necessary to solve the problem in order to improve
  learning stability and agent performance.

### Variable Length Observations

> It is possible to have observations be of variable size by using a `BufferSensor`,

Reviewer: ```suggestion
It is possible for agents to collect observations from a varying number of GameObjects by using a `BufferSensor`.
```

Reviewer 2: Suggestion on the suggestion - I don't think mentioning "GameObjects" is neecssary

You can add a `BufferSensor` to your Agent by adding a `BufferSensorComponent` to
its GameObject.
The `BufferSensor` can be useful in situations in which the Agent must pay

> attention to a varying number of entities. On the trainer side, the `BufferSensor`

Reviewer: Should we add an example e.g. varying number of tiles in the Sorter

Reviewer 2: I think "variable number of enemies" or "variable number of projectiles" is a better example.

> is processed using an attention module, more information about attention

Reviewer: ```suggestion
is processed using an attention module. More information about attention
```

mechanisms can be found [here](https://arxiv.org/abs/1706.03762). Training or
doing inference with variable length observations can be slower than using

> a flat vector observation, but can allow you to represent more complex problems

Reviewer: solve more complex problems?

Reviewer: ```suggestion
a flat vector observation. However, attention mechanism enable solving problems that require basic symbolic reasoning
```

such as our [Sorter environmentt](Learning-Environment-Examples.md#sorter).
Note that even though the `BufferSensor` can process a variable number of
entities, you still need to define a maximum number of entities. This is
because our network architecture requires to know what the shape of the

> observations will be. If less entities are processed than the maximum, the
> observation will be padded with zeros, but the trainer will actually ignore
> the padding.

Reviewer: ```suggestion
observations will be. If fewer entities are observed than the maximum, the
observation will be padded with zeros and the trainer will ignore
the padded observations.
```

> 

Reviewer: Maybe make a note that Attention layers are invariant to the order of variable entities so no preprocessing to properly 'order' obs is necessary

The `BufferSensor` constructor and Editor inspector have two arguments:
 - `Observation Size` : This is how many floats each entities will be
 represented with. This number is fixed and all entities must
 have the same representation. For example, if the entities you want to
 put into the `BufferSensor` have for relevant information position and
 speed, then the `Observation Size` should be 6 floats.
 - `Maximum Number of Entities` : This is the maximum number of entities
 the `BufferSensor` will be able to collect.

> 

Reviewer: is it obvious that you can add multiple buffer sensors?

Author: No, but we do not have an environment to show for it yet so I did not want to advertize it

Note that even though the `BufferSensor` can process a variable number of
entities, you still need to define a maximum number of entities. This is
because our network architecture requires to know what the shape of the
observations will be. If fewer entities are observed than the maximum, the
observation will be padded with zeros and the trainer will ignore
the padded observations. Note that attention layers are invariant to
the order of the entities, so there is no need to properly "order" the
entities before feeding them into the `BufferSensor`.

The `BufferSensor` constructor and Editor inspector have two arguments:
 - `Observation Size` : This is how many floats each entities will be
 represented with. This number is fixed and all entities must
 have the same representation. For example, if the entities you want to
 put into the `BufferSensor` have for relevant information position and
 speed, then the `Observation Size` should be 6 floats.
 - `Maximum Number of Entities` : This is the maximum number of entities
 the `BufferSensor` will be able to collect.

To add an entity's observations to the `BufferSensor`, you need to call the

> `BufferSensor.AppendObservation()` (or `BufferSensorComponent.AppendObservation()`)

Reviewer 2: I think this is a little misleading since you can't access the `BufferSensor` directly (at the moment)
